### ISSUE TITLE
Adding timestamp to cell

### ISSUE THREAD

**Issue by happiehappie** [jessesquires/JSQMessagesViewController#1551]
Hi again guys,

I need to add timestamp into chat bubble now. Like this 
![img_0272](https://cloud.githubusercontent.com/assets/6497131/14562036/fd9346fa-034c-11e6-911d-dfa851206d0a.jpg)
![img_0273](https://cloud.githubusercontent.com/assets/6497131/14562037/fd963202-034c-11e6-93f6-8dd02c63510c.jpg)
I referred to the documentation and in 

    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = super.collectionView(collectionView, cellForItemAtIndexPath: indexPath)
            as! JSQMessagesCollectionViewCell
        }

I referred to the documentation and tried adding my timestamp label into messageBubbleContainerView, also tried adding it to mediaView.

However, the label is still nowhere to be seen. (I gave it a frame and black color textColor and text just in case it's my carelessness). 
Also, even if I successfully added my label, how do I increase the height and width of my bubble?
Not sure if this has been answered/asked elsewhere but can I ask for advice on this?

**Comment by hemantkadam** [comment]
Its very simple @happiehappie all you need to do is subclass JSQMediaItem, write all your code in -(UIView *) mediaView, here you can create view with required CGSize and add it as subview to self.cachedVideoImageView/ self.cachedTextImageView/ self.cachedPhotoImageView. 

I've done this and trust me its very simple and will take less than 10 mins to get your required UI up and running. That's the best feature of JSQMessagesViewController, easy to customize ;)

Let me know if you need any help...

**Comment by jessesquires** [comment]
Thanks @hemantkadam !

**Comment by hemantkadam** [comment]
No problem @jessesquires this is least I can do to support JSQMessagesViewController  :+1:

**Comment by happiehappie** [comment]
Thanks for the reply. So I no longer dequeue my messages as JSQMessages but
media item instead?

**Comment by eliburke** [comment]
@happiehappie no, you still use JSQMessages. You need to create a new type of JSQMediaItem, call it HappiePhotoDateMediaItem or something like that. When you process your message data and have an image, you will create an object with the new type and set your image and date as properties. Internally you will modify the mediaView method, create a custom UIView, and add the date in a label, and add the image in a UIImageView. Add them as subviews to the custom view, and return it.

This is how all of the different chat data types are handled, and if you aren't sure how to proceed, you might want to take a look at the existing PhotoMedia item, and perhaps some of the various custom cells people have posted code for.

**Comment by Andrew05** [comment]
@hemantkadam,@eliburke, is the layout done programatically or with an xib file?

**Comment by hemantkadam** [comment]
@Andrew05 No, you don't need xib. I've done entire customizing programmatically.

**Comment by AashishSapkota** [comment]
@hemantkadam Can you please give an example of the entire customization programatically. Or, can we not give default value to **cell.celltoplabel.text** and **cell.cellbuttomlabel.text**?

**Comment by marcuschoong** [comment]
Hi, do you able to change the cell? i wish to do the same thing, possible you can provide the entire program as well?

**Comment by ikyh** [comment]
@hemantkadam   Hi. Can you please share the code for solution you mentioned above. I tried a lot but did not get any answer

**Comment by hemantkadam** [comment]
@ikyh @marcuschoong Cant share entire code but i have managed to put some kind of snippet, hope it helps to understand what and how you can do this. Code is just random but will give you idea of how you can do this. 
\#pragma mark - JSQMessageMediaData protocol
\- (UIView *)mediaView {
    if (self.cachedTextImageView == nil) {
        CGSize size = [self mediaViewDisplaySize];
        BOOL outgoing = self.appliesMediaViewMaskAsOutgoing;
        
        UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, size.width, size.height)];
        imageView.backgroundColor = colorBackground;
        imageView.clipsToBounds = YES;
        imageView.tag = _tag;

        UILabel *labelTime = [[UILabel alloc] initWithFrame:CGRectMake(X, Y, W, H)];
        labelTime.textAlignment = NSTextAlignmentLeft;
        labelTime.font = [labelTime.font fontWithSize:11];
        labelTime.textColor = [UIColor grayColor];
        labelTime.lineBreakMode = TEXT_LINE_BREAK;
        labelTime.numberOfLines = 0;
        labelTime.text = _time;
        
        [imageView addSubview:labelTime];
        
        [JSQMessagesMediaViewBubbleImageMasker applyBubbleImageMaskToMediaView:imageView isOutgoing:outgoing];
        self.cachedTextImageView = imageView;
    }
    return self.cachedTextImageView;
}

**Comment by marcuschoong** [comment]
Here is my way to add in the time stamp.

1. first add a uilabel on the JSQMessageCollectionViewCellIncoming.xib and assign to class "JSQMessageLabel"
<img width="727" alt="screen shot 2017-12-04 at 10 15 21 am" src="https://user-images.githubusercontent.com/26889594/33533509-86e8ca68-d8dc-11e7-97cb-119e13b9c155.png">
2. Bind the UIlabel at JSQMessageCollectionViewCell.m
<img width="792" alt="screen shot 2017-12-04 at 10 22 09 am" src="https://user-images.githubusercontent.com/26889594/33533593-1d012c98-d8dd-11e7-95b3-305547929029.png">
3. Finally you may assign uilabel.text at cellForItemAtIndexPath from ChatViewController.m
<img width="366" alt="screen shot 2017-12-04 at 10 24 52 am" src="https://user-images.githubusercontent.com/26889594/33533628-661498b6-d8dd-11e7-8977-5c7556fbbe51.png">

Hope this can help others ppl who looking on this feature.

**Comment by ikyh** [comment]
@marcuschoong Hi.. Thanks for replying. I am doing my project in swift 3. What if I want my own custom cell. Since I have to add buttons as well. I have copy pasted the `JSQMessagesCollectionViewCellIncoming` .xib and done some changes in that. I have also changed the subclass name to my subclass name. Though I have pasted the label on TextView I am getting timeStamp displayed at 'bubble top label'. I want in messagebubble.
